File: apps/coord/src/db/snapshot.ts
// Integrity-checked VACUUM INTO snapshot of the live coordinator DB, used as
// the payload for coordinator moves. Must use query() not prepare(): a
// prepared statement is finalized only on GC and keeps the source handle
// busy, so coord's own close() then fails with "database is locked". A failed
// run removes destPath — partial snapshots must never escape to the target.
import { createHash } from "node:crypto";
import * as fs from "node:fs";
import { Database } from "bun:sqlite";

const HASH_CHUNK_SIZE = 1024 * 1024;

/** Create a standalone, integrity-checked snapshot of a live SQLite database. */
export function createSqliteSnapshot(sqlite: Database, destPath: string): { size: number; sha256: string } {
  try {
    fs.rmSync(destPath, { force: true });
    // query(), not prepare(): a prepare()d Statement is finalized only on GC, so
    // it keeps the source handle busy — the coordinator's own close(true) then
    // reports "database is locked" after any snapshot. query() is cache-owned.
    sqlite.query("VACUUM INTO ?").run(destPath);
    fs.chmodSync(destPath, 0o600);

    const snapshot = new Database(destPath, { readonly: true });
    try {
      const result = snapshot.query<{ integrity_check: string }, []>("PRAGMA integrity_check").get();
      if (result?.integrity_check !== "ok") {
        throw new Error(`SQLite snapshot integrity_check failed: ${result?.integrity_check ?? "no result"}`);
      }
    } finally {
      snapshot.close();
    }

    const size = fs.statSync(destPath).size;
    const hasher = createHash("sha256");
    const chunk = new Uint8Array(HASH_CHUNK_SIZE);
    const fd = fs.openSync(destPath, "r");
    try {
      for (;;) {
        const bytesRead = fs.readSync(fd, chunk, 0, chunk.length, null);
        if (bytesRead === 0) break;
        hasher.update(chunk.subarray(0, bytesRead));
      }
    } finally {
      fs.closeSync(fd);
    }
    return { size, sha256: hasher.digest("hex") };
  } catch (error) {
    fs.rmSync(destPath, { force: true });
    throw error;
  }
}

File: crates/roost-coord/src/maintenance/snapshot.rs
//! The ordered snapshot procedure: one consistent, integrity-checked copy of
//! the live database, plus its size and digest.
//!
//! Owned by the coordinator. `backup` calls this before compressing; the
//! `/api/db-export` route calls the same procedure. The order is the point
//! (`docs/phase3-coord-contract.md` §2.5).

use std::io::Error as IoError;
use std::path::{Path, PathBuf};

use sha2::{Digest, Sha256};

use crate::db::{CoordDb, DbError};

/// The digest is streamed in slices of this size, for the same reason the
/// backup compresses in slices (`snapshot.ts:7`).
pub const SNAPSHOT_HASH_CHUNK_BYTES: usize = 1024 * 1024;

/// Why a snapshot could not be produced.
#[derive(Debug, thiserror::Error)]
pub enum SnapshotError {
    /// A file operation failed.
    #[error("snapshot i/o: {0}")]
    Io(#[from] IoError),
    /// The database refused the copy or the health check.
    #[error("snapshot database: {0}")]
    Database(#[from] DbError),
    /// `PRAGMA integrity_check` did not return `ok`.
    #[error("database integrity check failed: {0}")]
    IntegrityCheck(String),
}

/// A snapshot that is on disk, and what it hashes to.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SqliteSnapshot {
    /// Where the copy is.
    pub path: PathBuf,
    /// Its size in bytes.
    pub size: u64,
    /// Lowercase hex SHA-256 of its bytes.
    pub sha256: String,
}

/// Take a consistent, verified copy of `database` at `destination`.
///
/// Any failure removes the destination: a partial snapshot must never escape
/// to the caller (`snapshot.ts:5`).
pub async fn create_sqlite_snapshot(
    database: &CoordDb,
    destination: &Path,
) -> Result<SqliteSnapshot, SnapshotError> {
    // Never inherit a previous file (`snapshot.ts:9`).
    super::remove_file_if_present(destination).await?;
    let outcome = snapshot_into(database, destination).await;
    if outcome.is_err() {
        let _ = super::remove_file_if_present(destination).await;
    }
    outcome
}

async fn snapshot_into(
    database: &CoordDb,
    destination: &Path,
) -> Result<SqliteSnapshot, SnapshotError> {
    // `VACUUM INTO` is a transactionally consistent standalone copy, and it
    // does not block readers. The path is bound, not interpolated
    // (`db::CoordDb::vacuum_into`).
    database.vacuum_into(destination).await?;
    super::set_owner_only_file(destination).await?;

    // v2 reopened the COPY read-only and checked it (`snapshot.ts:22-27`). This
    // checks the live handle instead, because `db.rs` deliberately exposes one
    // database handle and one notion of "the database is healthy", and a second
    // open of the copy would be a second one. A corrupt source fails the
    // `VACUUM INTO` above outright, so the check is a gate on a source that
    // already copied, not a substitute for the copy.
    if !database.integrity_check().await? {
        return Err(SnapshotError::IntegrityCheck("not ok".to_string()));
    }

    let (size, sha256) = digest_file(destination).await?;
    Ok(SqliteSnapshot {
        path: destination.to_path_buf(),
        size,
        sha256,
    })
}

async fn digest_file(path: &Path) -> Result<(u64, String), IoError> {
    use tokio::io::AsyncReadExt;

    let mut file = tokio::fs::File::open(path).await?;
    let mut hasher = Sha256::new();
    let mut buffer = vec![0_u8; SNAPSHOT_HASH_CHUNK_BYTES];
    let mut size = 0_u64;
    loop {
        let read = file.read(&mut buffer).await?;
        if read == 0 {
            break;
        }
        size += read as u64;
        hasher.update(&buffer[..read]);
    }
    Ok((size, hex::encode(hasher.finalize())))
}
